File: crates/runtime-auth/src/layer/http.rs
/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use crate::{AuthVerdict, HttpAuth};
use axum::{
    body::Body,
    http::StatusCode,
    http::{Request, Response},
};
use futures::future::BoxFuture;
use std::sync::Arc;
use std::task::{Context, Poll};
use tower::{Layer, Service};

#[derive(Clone)]
pub struct AuthLayer {
    auth_verifier: Arc<dyn HttpAuth + Send + Sync>,
}

impl AuthLayer {
    pub fn new(auth_verifier: Arc<dyn HttpAuth + Send + Sync>) -> Self {
        Self { auth_verifier }
    }
}

impl<S> Layer<S> for AuthLayer {
    type Service = AuthMiddleware<S>;

    fn layer(&self, inner: S) -> Self::Service {
        AuthMiddleware {
            inner,
            auth_verifier: Arc::clone(&self.auth_verifier),
        }
    }
}

#[derive(Clone)]
pub struct AuthMiddleware<S> {
    inner: S,
    auth_verifier: Arc<dyn HttpAuth + Send + Sync>,
}

impl<S> Service<Request<Body>> for AuthMiddleware<S>
where
    S: Service<Request<Body>, Response = Response<Body>> + Clone + Send + 'static,
    S::Future: Send + 'static,
{
    type Response = S::Response;
    type Error = S::Error;
    type Future = BoxFuture<'static, Result<Response<Body>, Self::Error>>;

    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        self.inner.poll_ready(cx)
    }

    fn call(&mut self, req: Request<Body>) -> Self::Future {
        let auth_verifier = Arc::clone(&self.auth_verifier);
        let mut inner = self.inner.clone();

        Box::pin(async move {
            let (parts, body) = req.into_parts();

            match auth_verifier.http_verify(&parts) {
                Ok(AuthVerdict::Allow(_)) => inner.call(Request::from_parts(parts, body)).await,
                Ok(AuthVerdict::Deny) => Ok(unauthorized_response()),
                Err(e) => {
                    tracing::error!("{e}");
                    Ok(internal_server_error_response())
                }
            }
        })
    }
}

fn unauthorized_response() -> Response<Body> {
    match Response::builder()
        .status(StatusCode::UNAUTHORIZED)
        .body(Body::from("Unauthorized"))
    {
        Ok(response) => response,
        Err(e) => panic!("Failed to build unauthorized response: {e}"),
    }
}

fn internal_server_error_response() -> Response<Body> {
    match Response::builder()
        .status(StatusCode::INTERNAL_SERVER_ERROR)
        .body(Body::from("Internal Server Error"))
    {
        Ok(response) => response,
        Err(e) => panic!("Failed to build internal server error response: {e}"),
    }
}

File: test/harness/http_cache_ratelimit/loadgen/harness/http.py
# Copyright 2024-2026 The Spice.ai OSS Authors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Minimal urllib-based JSON / SQL client for the harness.

Kept dependency-free (standard library only) so the load generators need no
third-party HTTP client. ``sql`` never raises: any transport or HTTP error is
returned as a ``(status, body)`` pair so the caller can record it as a sample
rather than aborting the run.
"""

from __future__ import annotations

import json
import urllib.error
import urllib.request
from typing import Any


def get_json(url: str, timeout: float = 5.0) -> dict[str, Any]:
    """GET ``url`` and decode a JSON object. Raises on transport/HTTP error."""
    req = urllib.request.Request(url, method="GET")
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        return json.loads(resp.read().decode())


def post_json(url: str, payload: dict[str, Any], timeout: float = 5.0) -> Any:
    """POST ``payload`` as JSON and decode the JSON response. Raises on error."""
    body = json.dumps(payload).encode()
    req = urllib.request.Request(
        url, data=body, headers={"Content-Type": "application/json"}, method="POST"
    )
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        return json.loads(resp.read().decode())


def sql(url: str, query: str, timeout: float = 5.0) -> tuple[int, Any]:
    """POST a SQL string to spiced's ``/v1/sql`` and return ``(status, body)``.

    Never raises: an HTTP error yields ``(code, text)`` and any other transport
    error yields ``(0, "<ErrorType>: <message>")`` so the caller records it as a
    sample.
    """
    req = urllib.request.Request(
        url, data=query.encode(), headers={"Content-Type": "text/plain"}, method="POST"
    )
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            return resp.status, json.loads(resp.read().decode())
    except urllib.error.HTTPError as e:
        return e.code, e.read().decode(errors="replace")
    except Exception as e:  # noqa: BLE001 - surface any transport error as a sample
        return 0, f"{type(e).__name__}: {e}"
